### ISSUE TITLE
[DependencyInjection] PhpDumper, abstract service and factory.

### ISSUE THREAD

**Issue by fbourigault** [symfony/symfony#17353]
While using Symfony 2.7.6 (composer update won't let me update to 2.7.8 even with a ^2.7.0 dependency :/) and [csa/guzzle-bundle](https://github.com/csarrazi/CsaGuzzleBundle) 2.0.4 and [guzzlehttp/guzzle](https://github.com/guzzle/guzzle) 6.1.1 I encountered a bug that lead to generate the following code in `appDevDebugProjectContainer.php` (I removed some sensitive data):
```php
    /**
     * Gets the 'csa_guzzle.client.my_client' service.
     *
     * This service is shared.
     * This method always returns the same instance of the service.
     *
     * @return \GuzzleHttp\Client A GuzzleHttp\Client instance.
     */
    protected function getCsaGuzzle_Client_MyClientService()
    {
        $a = new \UL();
        $a->push($this->get('csa_guzzle.middleware.stopwatch'), 'stopwatch');
        $a->push($this->get('csa_guzzle.middleware.logger'), 'logger');
        $a->push($this->get('csa_guzzle.middleware.history'), 'history');

        return $this->services['csa_guzzle.client.my_client'] = new \GuzzleHttp\Client(array('base_uri' => 'https://...', 'headers' => array('Authorization' => 'Bearer ...', 'Content-Type' => '...'), 'on_stats' => array(0 => $this->get('csa_guzzle.data_collector.guzzle'), 1 => 'addStats'), 'handler' => $a));
    }
```

The service is defined from an abstract service using a factory:
https://github.com/csarrazi/CsaGuzzleBundle/blob/v2.0.4/src/Resources/config/middleware.xml#L9
and decorated in a compiler pass:
https://github.com/csarrazi/CsaGuzzleBundle/blob/v2.0.4/src/DependencyInjection/CompilerPass/MiddlewarePass.php#L117

I did some debug in PhpDumper and notice that in:
https://github.com/symfony/symfony/blob/v2.7.6/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L752
the `$class` value returned is NULL and:
https://github.com/symfony/symfony/blob/v2.7.6/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L800
makes it look like `$a = new \UL()`.

I don't know if it's a csa/guzzle-bundle misuse of DependencyInjection or a bug in the PhpDumper class.

**Comment by fbourigault** [comment]
I upgraded to symfony/symfony 2.8.1 and the code generated is fine.

**Comment by fbourigault** [comment]
Looks like it was fixed by d3a4a77 that depends on feature introduced in #14984.

I will try to upgrade my project to 2.8

**Comment by fbourigault** [comment]
It's working using 2.8.1/2.8.2

**Comment by jakzal** [comment]
The difference is [2.7 ignores abstract definitions](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php#L46) when resolving definition templates, while [2.8 does not](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php#L42). The change was introduced in #15096.

**Comment by xabbuh** [comment]
Closing here as the issue was in the csarrazi/CsaGuzzleBundle package and was fixed in csarrazi/CsaGuzzleBundle#120.

**Comment by jakzal** [comment]
Just to double check... ;)

@nicolas-grekas do you remember why we no longer ignore abstract services in the ResolveDefinitionTemplatesPass (see my previous comment)?
